examples/event_loop/src/event_loop_example1.cpp

```diff
@@ -0,0 +1,110 @@
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+#include <klepsydra/high_performance/event_loop_middleware_provider.h>
+#include <klepsydra/vision_ocv/file_image_stream_service.h>
+#include <klepsydra/vision_ocv/image_data_factory.h>
+#include <klepsydra/core/publisher.h>
+#include <klepsydra/core/subscriber.h>
+#include "spdlog/spdlog.h"
+#include "thread_id_helper.h"
+
+int main(int argc, char *argv[]) {
+  if (argc < 2) {
+    spdlog::error("Input folder not provided.");
+    return -1;
+  } else if (argc < 3) {
+    spdlog::error("Output folder not provided.");
+    return -1;
+  } else if (argc >= 4 && !std::strcmp(argv[3], "debug")) {
+    spdlog::set_level(spdlog::level::debug);
+  } 
+
+  const std::string inputFolderName = argv[1], outputFolderName = argv[2];
+  const int width = 500, height = 500;
+  const int totalCameras = 3, readTimes=6;
+
+  kpsr::vision_ocv::ImageDataFactory imageCreator(width, height, cv::IMREAD_REDUCED_GRAYSCALE_2, "camera_link");
+  kpsr::high_performance::EventLoopMiddlewareProvider<16> eventLoopProvider(nullptr);
+  std::vector<kpsr::Subscriber<kpsr::vision_ocv::ImageData>*> subscribers;
+  std::vector<kpsr::Publisher<kpsr::vision_ocv::ImageData>*> publishers;
+  std::vector<std::unique_ptr<kpsr::vision_ocv::FileImageStreamingService>> readFromDiskServices;
+  std::vector<std::thread> threads;
+  std::array<int, totalCameras> frameCounters{};
+
+  eventLoopProvider.start();
+  
+  std::string prefixChannel = "channel_";
+
+  // subscriber creation and listener registration.
+  for (int i = 0; i < totalCameras; i++) {
+    std::string channelName = prefixChannel+std::to_string(i);
+    subscribers.push_back(eventLoopProvider.getSubscriber<kpsr::vision_ocv::ImageData>(channelName));
+    const std::string outputFolderPrefix = outputFolderName + "/camera_id_"+std::to_string(i)+"_frame";
+    subscribers[i]->registerListener(channelName, [i, &frameCounters, outputFolderPrefix](const kpsr::vision_ocv::ImageData &event) {
+            std::string imgName = outputFolderPrefix + std::to_string(frameCounters[i]++) + ".jpg"; 
+            cv::Mat outputImage;
+            cvtColor(event.img, outputImage,cv::COLOR_BGR2GRAY);
+            bool statusWrite = cv::imwrite(imgName, outputImage);
+            spdlog::debug("Status imwrite: {}.", statusWrite);
+            spdlog::info("Listener. Thread-id: {}", getPrintableThreadId());
+    });
+  }
+
+  // publisher creation
+  for (int i = 0; i < totalCameras; i++) {
+    publishers.push_back(eventLoopProvider.getPublisher<kpsr::vision_ocv::ImageData>(prefixChannel+std::to_string(i),
+     0, imageCreator.initializerFunction, imageCreator.eventClonerFunction));
+  }
+
+  // Services creation
+  for (int i = 0; i < totalCameras; i++) {
+    readFromDiskServices.emplace_back(new kpsr::vision_ocv::FileImageStreamingService(
```

Comment:
Does this need to be vector of unique_ptrs ? Can you not just have a vector of FileImageStreamingService?